File: assets/js/game.js
import React from 'react';
import Board from './grid.js';
import socket from './socket';
import run_demo from "./demo";

var channel;

function init() {
	let root = document.getElementById('root');
	if(!root) {
		return;
	}

	let channel = socket.channel("game:" + name, {observer: othello_cfg.observer});
	run_demo(root, channel);

}

$(init);

function clickEvent(i) {
	const clickLog = this.state.clickLog.slice(0, this.state.clickCount + 1);
	const present = clickLog[this.state.clickCount];

	if (this.gameResult(present.xCount, present.oCount) || present.tiles[i]) {
		return;
	}

	const changedTiles = this.changeTiles(present.tiles, i, this.state.xTurn);

	if (changedTiles === null) {
		return;
	}

	const xCount = changedTiles.reduce((acc, present) => { return present === 'X' ? acc + 1 : acc }, 0);
	const oCount = changedTiles.reduce((acc, present) => { return present === 'O' ? acc + 1 : acc }, 0);

	let skipTurn = this.possibleMoves(!this.state.xTurn, changedTiles).length > 0 ? !this.state.xTurn : this.state.xTurn

	this.setState({
		clickLog: clickLog.concat([{
			tiles: changedTiles,
			xCount: xCount,
			oCount: oCount,
			wasXTurn: skipTurn
		}]),
		clickCount: clickLog.length,
		xTurn: skipTurn,
	});
}

function gameResult(xCount, oCount) {
	return (xCount + oCount < 64) ? null : (xCount === oCount) ? 'XO' : (xCount > oCount ? 'X' : 'O');
}

function changeTiles(tiles, location, xTurn) {
	let changedGrid = null;

	let [beginX, beginY] = [location % 8, (location - location % 8) / 8];

	if (tiles[location] !== null) {
		return null;
	}

	[1, 7, 8, 9, -1, -7, -8, -9].forEach((weight) => {
		let changedTiles = changedGrid ? changedGrid.slice() : tiles.slice();
		let oneOrMoreChanged = false;
		let [previousXLocation, previousYLocation] = [beginX, beginY];

		for (let y = location + weight; y < 64; y = y + weight) {

			let [xLocation, yLocation] = [y % 8, (y - y % 8) / 8];

			if (Math.abs(previousXLocation - xLocation) > 1 || Math.abs(previousYLocation - yLocation) > 1) {
				break;
			}

			if (changedTiles[y] === (!xTurn ? 'X' : 'O')) {
				changedTiles[y] = xTurn ? 'X' : 'O';
				oneOrMoreChanged = true;
				[previousXLocation, previousYLocation] = [xLocation, yLocation];
				continue;
			}

			else if ((changedTiles[y] === (xTurn ? 'X' : 'O')) && oneOrMoreChanged) {
				changedTiles[location] = xTurn ? 'X' : 'O';
				changedGrid = changedTiles.slice();
			}
			break;
		}
	});

	return changedGrid;
}

function possibleMoves(color, tiles) {
	return tiles
		.map((value, index) => { return this.changeTiles(tiles, index, color) ? index : null; })
		.filter((element) => { return element !== null; });
}

function gotView(msg) {
	this.setState(msg.game);
}

export default class Game extends React.Component {

	constructor(props) {

		super(props);

		this.channel = socket.channel("game:" + othello_cfg.game, {user: othello_cfg.user, host: othello_cfg.host, observer: othello_cfg.observer});

		const tilesInitial = Array(64).fill(null);
		[tilesInitial[8 * 3 + 3], tilesInitial[8 * 3 + 4], tilesInitial[8 * 4 + 4], tilesInitial[8 * 4 + 3]] = ['X', 'O', 'X', 'O'];

		this.state = {
			clickLog: [{
				tiles: tilesInitial,
				xCount: 2,
				oCount: 2,
				wasXTurn: true
			}],
			clickCount: 0,
			xTurn: true,

		}

		this.channel.join()
			.receive("ok", resp => {console.log("Joined Successfully", resp)})
			.receive("error", resp => {console.log("Unable to Join", resp);});

		const tilesInitial1 = Array(64).fill(null);
		[tilesInitial1[8 * 3 + 3], tilesInitial1[8 * 3 + 4], tilesInitial1[8 * 4 + 4], tilesInitial1[8 * 4 + 3]] = ['X', 'O', 'X', 'O'];

		const nextSquares = Array(64).fill(null);
		nextSquares[20] = 'X';
		nextSquares[27] = 'X';
		nextSquares[28] = 'X';
		nextSquares[35] = 'O';
		nextSquares[36] = 'X';

		this.channel.on("clickEvent", state => {

			this.setState({
				clickLog: state["clickLog"],
				xTurn: state["xTurn"],
				clickCount: state["clickCount"]});

		});

		this.channel.on("reset", state => {

			this.setState({
				clickLog: state["clickLog"],
				xTurn: state["xTurn"],
				clickCount: state["clickCount"]});

		});

	}

	gotView(msg) {
		this.setState(msg.game);
	}

	gameResult(xCount, oCount) {
		return (xCount + oCount < 64) ? null : (xCount === oCount) ? 'XO' : (xCount > oCount ? 'X' : 'O');
	}

	changeTiles(tiles, location, xTurn) {
		let changedGrid = null;

		let [beginX, beginY] = [location % 8, (location - location % 8) / 8];

		if (tiles[location] !== null) {
			return null;
		}

		[1, 7, 8, 9, -1, -7, -8, -9].forEach((weight) => {
			let changedTiles = changedGrid ? changedGrid.slice() : tiles.slice();
			let oneOrMoreChanged = false;
			let [previousXLocation, previousYLocation] = [beginX, beginY];

			for (let y = location + weight; y < 64; y = y + weight) {

				let [xLocation, yLocation] = [y % 8, (y - y % 8) / 8];

				if (Math.abs(previousXLocation - xLocation) > 1 || Math.abs(previousYLocation - yLocation) > 1) {
					break;
				}

				if (changedTiles[y] === (!xTurn ? 'X' : 'O')) {
					changedTiles[y] = xTurn ? 'X' : 'O';
					oneOrMoreChanged = true;
					[previousXLocation, previousYLocation] = [xLocation, yLocation];
					continue;
				}

				else if ((changedTiles[y] === (xTurn ? 'X' : 'O')) && oneOrMoreChanged) {
					changedTiles[location] = xTurn ? 'X' : 'O';
					changedGrid = changedTiles.slice();
				}
				break;
			}
		});

		return changedGrid;
	}

	possibleMoves(color, tiles) {
		return tiles
			.map((value, index) => { return this.changeTiles(tiles, index, color) ? index : null; })
			.filter((element) => { return element !== null; });
	}

	render() {

		const clickLog = this.state.clickLog.slice();
		const present = clickLog[this.state.clickCount];

		const xTurn = this.state.xTurn;

		let winner = this.gameResult(present.xCount, present.oCount);

		let movesPossible = this.possibleMoves(present.wasXTurn, present.tiles);
		let availableMovesOpposite = this.possibleMoves(!present.wasXTurn, present.tiles);

		if ((movesPossible.length === 0) && (availableMovesOpposite.length === 0))
		{
			winner = present.xCount === present.oCount ? 'XO' : present.xCount > present.oCount ? 'X' : 'O';
		}

		let gameStatus =
			winner ?
				(winner === 'XO') ? 'It\'s a draw' : 'The winner is ' + (winner === 'X' ? 'Host(White)' : 'black(Other Player)') :
				[this.state.xTurn ? 'Hosts(Whites) turn' : 'Other Players(Blacks) turn', ' with ', movesPossible.length, ' available moves'].join('');

		return (

			<div>
			<div className="rules">
			<h1 align="center"> Rules </h1>
			<ul>
			<li>
			The player can reverse opponent's disc when the newly placed disc surrounds the opponent's disc in a straight line in horizontal, vertical or diagonal manner.
			</li>
			<br />
			<li>
			Legal move is only considered when the disc could be placed where it outflanks atleast one of the opponent's disc. Otherwise the player has to pass his turn and wait till he can make a legal move.
			</li>
			<br />
			<li>
			The player has to play his move and cannot forfeit until a legal move is possible even if forfeiting the move could be advantageous to the player.
			</li>
			<br />
			<li>
			The player can flip all possible discs of the opponent that are outflanked by the current player in a single move in all the direction.
			</li>
			</ul>
			</div>
			<div className="game">
			<div className="top">{gameStatus}&nbsp;{winner ? <button className="reset" onClick={() => this.resetClickCondition()}>Play again</button> : ''}</div>


			<div className="game-left-side">

				<div className="game-grid">

					  <Board tiles={present.tiles} movesPossible={movesPossible} onClick={(index) => this.clickCondition(index)}/>

				</div>

				<div className="resetDiv"><button id="reset" className="reset"onClick={() => this.resetClickCondition()}>Reset Game</button></div>

				<div></div>
			</div>
				<div className="info">

					<div> User: <b> {othello_cfg.user} </b> </div>
					<div> Name:<b>  {othello_cfg.game} </b> </div>
					<br />
					<br />
					<div>Host(White) count: <b> {present.xCount} </b> </div>

					<div>Other Player(Black) count: <b> {present.oCount} </b> </div>
					<br />

				</div>
				<div>
				</div>
			</div>
			</div>
		);
	}

	resetClickCondition() {
		const isObserver = othello_cfg.observer;
		if(isObserver=="false") {
			this.channel.push("reset",{state: this.state}).receive("ok", resp => {this.gotView(resp)})
		}

	}

	clickCondition(index) {
		const host = othello_cfg.host;
		const isObserver = othello_cfg.observer;

		if((host == "true") && (this.state.xTurn) && !(isObserver=="true")) {
			this.channel.push("clickEvent",{index: index, state: this.state}).receive("ok", resp => {this.gotView(resp)});
		}
		else if((host == "false") && (!this.state.xTurn) && !(isObserver=="true")) {
			this.channel.push("clickEvent",{index: index, state: this.state}).receive("ok", resp => {this.gotView(resp)});
		}
	}
}

File: lib/othello/game.ex

defmodule Othello.Game do

  def start_link do
    Agent.start_link(fn -> %{} end, name: __MODULE__)
  end

  def put(gname, game) do
    Agent.update(__MODULE__, &Map.put(&1, gname, game))
    game
  end

  def get(gname) do
    Agent.get(__MODULE__, &Map.get(&1, gname))
  end

  def join(gname, user, observer) do
    grid = List.duplicate(nil, 64) |>
    List.replace_at(27, 'X') |>
    List.replace_at(28, 'O') |>
    List.replace_at(36, 'X') |>
    List.replace_at(35, 'O')

    game = get(gname)

    if game do
      game
    else
      game = %{name: gname, observer: observer, host: user, clickLog: %{"tiles" => grid, "xCount" => 2, "oCount" => 2, "wasXTurn" => true}, clickCount: 0, clickCount: 0}
      put(gname, game)
    end

  end

  def reset(state) do
    clickLog = Map.get(state, "clickLog")
    history0 = Enum.fetch!(clickLog, 0)
    wasXTurn = Map.get(history0, "wasXTurn")

    %{
      clickLog: Enum.slice(clickLog, 0, 1),
      clickCount: 0,
      xTurn: wasXTurn
    }
  end

  def gameResult(xCount, oCount) do
    if(xCount+oCount<64) do
      nil
    else
      if(xCount === oCount) do
        "XO"
      else
        if(xCount > oCount) do
          "X"
        else
          "O"
        end
      end
    end
  end

  def forfunc(y, previousXLocation, previousYLocation, xTurn, changedGrid, changedTiles, oneOrMoreChanged, location, weight) do

    xLocation = rem(y, 8)
    yLocation = (y - rem(y,8)) / 8

    terminatingCond1 = abs(previousXLocation - xLocation) > 1 || abs(previousYLocation - yLocation) > 1

    if(terminatingCond1) do
      changedGrid
    else
      if(!xTurn) do
        toMatch = "X"
      else
        toMatch = "O"
      end

      if(xTurn) do
        reverseLetter = "X"
      else
        reverseLetter = "O"
      end

      if(Enum.fetch!(changedTiles, y) === toMatch) do

        changedTiles = Enum.to_list(changedTiles)
        changedTiles = List.replace_at(changedTiles, y, reverseLetter)
        oneOrMoreChanged = true
        previousXLocation = xLocation
        previousYLocation = yLocation

        y = y + weight
        if (y<64) do
          changedGrid = forfunc(y, previousXLocation, previousYLocation, xTurn, changedGrid, changedTiles, oneOrMoreChanged, location, weight)
        else
          changedGrid
        end

      else
        if((Enum.fetch!(changedTiles,y) === reverseLetter) && oneOrMoreChanged) do

          changedTiles = Enum.to_list(changedTiles)
          changedTiles = List.replace_at(changedTiles, location, reverseLetter)
          changedGrid = Enum.slice(changedTiles, 0, Enum.count(changedTiles))
          changedGrid
        else
          changedGrid
        end
        changedGrid
      end
    end

  end

  def forEachfunc(weight, changedGrid, beginX, beginY, tiles, location, xTurn) do

    if (changedGrid) do
      changedTiles = Enum.slice(changedGrid, 0, Enum.count(changedGrid))
    else
      changedTiles = Enum.slice(tiles, 0, Enum.count(tiles))
    end

		oneOrMoreChanged = false
		previousXLocation = beginX
    previousYLocation = beginY

    if (location+weight<64) do
      y = location+weight
    else
      y = location
    end

    changedGrid = forfunc(y, previousXLocation, previousYLocation, xTurn, changedGrid, changedTiles, oneOrMoreChanged, location, weight)

    changedGrid
  end

  def changeTiles(tiles, location, xTurn) do
    changedGrid = nil

		beginX = rem(location, 8)
    beginY = (location - rem(location,8)) / 8

		if (Enum.fetch!(tiles, location) != nil) do
		   nil
		else

      changedGrid1 = forEachfunc(1, changedGrid, beginX, beginY, tiles, location, xTurn)
      changedGrid2 = forEachfunc(7, changedGrid1, beginX, beginY, tiles, location, xTurn)
      changedGrid3 = forEachfunc(8, changedGrid2, beginX, beginY, tiles, location, xTurn)
      changedGrid4 = forEachfunc(9, changedGrid3, beginX, beginY, tiles, location, xTurn)
      changedGrid5 = forEachfunc(-1, changedGrid4, beginX, beginY, tiles, location, xTurn)
      changedGrid6 = forEachfunc(-7, changedGrid5, beginX, beginY, tiles, location, xTurn)
      changedGrid7 = forEachfunc(-8, changedGrid6, beginX, beginY, tiles, location, xTurn)
      changedGrid8 = forEachfunc(-9, changedGrid7, beginX, beginY, tiles, location, xTurn)
      changedGrid8

    end
  end

  def possibleMoves(color, tiles) do
    squares0 = tiles |> Enum.with_index(0) |> Enum.map(fn {k,v}->{v,k} end) |> Map.new
    squaress = Enum.to_list(squares0)
    squares1 = Enum.map(squares0, fn{index,value} -> mapFunc(color, tiles, index) end)
    squares2 =
      Enum.filter(squares1, fn(element) -> element == nil end)
    squares2

  end

  def mapFunc(color, tiles, index) do
    if (changeTiles(tiles, index, color) == nil) do
      nil
    else
      index
    end
  end

  def reduceFunc1(acc, present) do
    if (present === "X") do
      acc = acc + 1
    else
      acc = acc
    end
  end

  def reduceFunc2(acc, present) do
    if (present === "O") do
      acc = acc + 1
    else
      acc = acc
    end
  end

  def clickEvent(index, state) do
    clickLog = Map.get(state, "clickLog")
    clickCount = Map.get(state, "clickCount")
    clickLog = Enum.slice(clickLog, 0, clickCount+1)
    present = Enum.fetch!(clickLog, clickCount)

    xCount = Map.get(present, "xCount")
    oCount = Map.get(present, "oCount")
    tiles = Map.get(present, "tiles")
    tile = Enum.fetch!(tiles, index)

    if (gameResult(xCount, oCount) || tile) do
      nil
    else
      changedTiles = changeTiles(Map.get(present, "tiles"), index, Map.get(state, "xTurn"))

      if(changedTiles == nil) do
        nil
      else

        xCount = Enum.reduce(changedTiles, 0, fn(present,acc) -> reduceFunc1(acc,present) end)
        oCount = Enum.reduce(changedTiles, 0, fn(present,acc) -> reduceFunc2(acc,present) end)

        list = possibleMoves(!Map.get(state,"xTurn"), changedTiles)

        if(length(list) > 0) do
          skipTurn = !Map.get(state, "xTurn")
        else
          skipTurn = Map.get(state, "xTurn")
        end

        clickLog = clickLog ++ [(%{"tiles" => changedTiles, "xCount" => xCount, "oCount" => oCount, "wasXTurn" => skipTurn})]

        state = Map.put(state, "clickLog", clickLog)
        state = Map.put(state, "clickCount", length(clickLog)-1)
        state = Map.put(state, "xTurn", skipTurn)

        %{
          clickLog: Map.get(state, "clickLog"),
          clickCount: Map.get(state, "clickCount"),
          xTurn: Map.get(state, "xTurn")
        }
      end
		end
  end
end
